### PR TITLE
Explicit keys

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 
implements: [KLIP-29](#5530)

fixes: #5303
fixes: #4678

This change sees ksqlDB no longer adding an implicit `ROWKEY STRING` key column to created streams or primary key column to created tables when no key column is explicitly provided in the `CREATE` statement.

BREAKING CHANGE

`CREATE TABLE` statements will now fail if not `PRIMARY KEY` column is provided.

For example, a statement such as:

```sql
CREATE TABLE FOO (name STRING) WITH (kafka_topic='foo', value_format='json');
```

Will need to be updated to include the definition of the PRIMARY KEY, e.g.

```sql
CREATE TABLE FOO (ID STRING PRIMARY KEY, name STRING) WITH (kafka_topic='foo', value_format='json');
```

If using schema inference, i.e. loading the value columns of the topic from the Schema Registry, the primary key can be provided as a partial schema, e.g.

```sql
-- FOO will have value columns loaded from the Schema Registry
CREATE TABLE FOO (ID INT PRIMARY KEY) WITH (kafka_topic='foo', value_format='avro');
```

`CREATE STREAM` statements that do not define a `KEY` column will no longer have an implicit `ROWKEY` key column.

For example:

```sql
CREATE STREAM BAR (NAME STRING) WITH (...);
```

The above statement would previously have resulted in a stream with two columns: `ROWKEY STRING KEY` and `NAME STRING`.
With this change the above statement will result in a stream with only the `NAME STRING` column.

Streams will no KEY column will be serialized to Kafka topics with a `null` key.

### Testing done 
usual

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json [hunk 1]
```diff
@@ -859,6 +859,52 @@
...
+      }
+    },
+    {
+      "name": "stream without key column",
+      "statements": [
+        "CREATE STREAM INPUT (ID INT, F0 INT) WITH (kafka_topic='input', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT;"
+      ],
+      "inputs": [
+        {"topic": "input", "key": null, "value": {"ID": 1, "F0": 2}}
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json:897]:
```suggestion
        {"topic": "input", "key": null, "value": {"ID": 1, "F0": 2}},
        {"topic": "input", "key": "foo", "value": {"ID": 1, "F0": 2}}
```
let's also test what happens when there is a key in the message but one isn't declared

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/joins.json [hunk 2]
```diff
@@ -1966,6 +1966,26 @@
       "outputs": [
         {"topic": "OUTPUT", "key": "user_0", "value": {"IMPRESSION_ID": 24, "URL": "urlA"}, "timestamp":  12}
       ]
+    },
+    {
+      "name": "streams with no key columns",
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/joins.json:1971]: dont want to be too picky, but let's also add a stream no-key --> table join

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json (hunk 1 above)
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json:897]: sure.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/joins.json (hunk 2 above)
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/joins.json]: done

**Pull request merged**